### PR TITLE
Implement jade sea

### PR DESCRIPTION
**Pull request by EvadLhorg**

### CODE REVIEW

File: server/game/player.js [hunk 1]
```diff
@@ -159,7 +161,18 @@ class Player extends Spectator {
     }
 
     getNumberOfUsedPlots() {
-        return this.plotDiscard.length + this.usedPlotsModifier;
+        return this.plotDiscard.length + this.usedPlotsModifier + this.usedPlotsModifierByTrait.getValues().reduce((sum, entry) => sum + this.usedPlotsModifierByTrait.getCountForReference(entry), 0);
```
Comment (ystros) [inline server/game/player.js:164]: Just to confirm this is intentional, this means "considered to have 1 additional <trait> plot in their used pile" counts toward the total number of plots in the used pile, not just when counting by that trait?

File: server/game/player.js [hunk 2]
```diff
@@ -213,6 +226,20 @@ class Player extends Spectator {
...
+            for(let i = 0; i < value; i++) {
+                this.usedPlotsModifierByTrait.add(trait);
+            }
+        } else {
+            for(let i = 0; i < value * -1; i++) {
+                this.usedPlotsModifierByTrait.remove(trait);
+            }
+        }
+        
+        this.game.raiseEvent('onUsedPlotsModified', { player: this });
```
Comment (ystros) [inline server/game/player.js:240]: This event isn't used anywhere. Were you finding that effects weren't being recalculated when adding / removing Gulltown from play? If so, I'd recommend adding a TODO comment here to fix those calculations. If not, we should get rid of it.

File: server/game/cards/13.1-AtG/AtTheGates.js [hunk 3]
```diff
@@ -16,11 +16,7 @@ class AtTheGates extends PlotCard {
     }
 
     hasUsedCityPlot(player) {
-        return this.game.allCards.some(card => (
-            card.controller === player &&
-            card.location === 'revealed plots' &&
-            card.hasTrait('City')
-        ));
+        return player.getNumberOfUsedPlotsByTrait('City') > 0;
```
Comment (ystros) [inline server/game/cards/13.1-AtG/AtTheGates.js:19]: 🥳

File: server/game/cards/19-JS/Brimstone.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            match: card => card.getType() === 'character' && card.controller === this.controller && card.hasTrait('Sand Snake'),
+            effect: ability.effects.modifyStrength(1)
+        });
+        
+        this.action({
+            title: 'Return Sand Snake',
+            phase: 'challenge',
+            cost: [
+                ability.costs.kneelSelf(),
+                ability.costs.returnToHand(card => card.getType() === 'character' && card.controller === this.controller && card.hasTrait('Sand Snake'))
```
Comment (ystros) [inline server/game/cards/19-JS/Brimstone.js:15]: Controller check should be unnecessary here - the cost system should be enforcing this.

File: server/game/cards/19-JS/Brimstone.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            ],
+            handler: context => {
+                this.game.addGold(this.controller, 1);
```
Comment (ystros) [inline server/game/cards/19-JS/Brimstone.js:18]: This and the message should be `context.player`.

File: server/game/cards/19-JS/CityFestival.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const PlotCard = require('../../plotcard.js');
+
+class CityFestival extends PlotCard {
+    setupCardAbilities() {
+        this.whenRevealed({
+            handler: context => {
+                if(context.player.canGainGold()) {
+                    let numGold = context.player.getNumberOfUsedPlotsByTrait('City') >= 2 ? 6 : 3;
+                    let gold = this.game.addGold(context.player, numGold);
+                    this.game.addMessage('{0} uses {1} to gain {2} gold', this.controller, this, gold);
```
Comment (ystros) [inline server/game/cards/19-JS/CityFestival.js:10]: `context.player`

File: server/game/cards/19-JS/Harlaw.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+const DrawCard = require('../../drawcard.js');
+
+class Harlaw extends DrawCard {
+    setupCardAbilities(ability) {
+        this.persistentEffect({
+            match: (card) => card === card.controller.activePlot,
+            targetController: 'opponent',
+            effect: ability.effects.modifyReserve(-1)
```
Comment (ystros) [inline server/game/cards/19-JS/Harlaw.js:8]: With all of the potential reserve reducers now, I had to double check that we don't allow it to go below 0. 😅

File: server/game/cards/19-JS/KingsLandingSmuggler.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        this.interrupt({
+            when: {
+                onCardLeftPlay: event => event.card === this 
+            },
+            target: {
+                activePromptTitle: 'Select a card without gold',
+                cardCondition: card => card.location === 'play area' && !card.hasToken(Tokens.gold)
+            },
+            handler: (context) => {
+                this.game.addMessage('{0} uses {1} to place a gold from the treasury on {2}', this.controller, this, context.target);
```
Comment (ystros) [inline server/game/cards/19-JS/KingsLandingSmuggler.js:16]: `context.player`

File: server/game/cards/19-JS/KingsLandingSmuggler.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+                this.game.resolveGameAction(
+                    GameActions.placeToken(() => ({ card: context.target, token: Tokens.gold, source: this })),
```
Comment (ystros) [inline server/game/cards/19-JS/KingsLandingSmuggler.js:18]: The place token action doesn't have a `source` parameter, just `card`, `token`, and optionally `amount`.

File: server/game/cards/19-JS/LeftHandLew.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    setupCardAbilities(ability) {
+        this.action({
+            title: 'Stand character',
+            cost: ability.costs.kneelSelf(),
+            target: {
+                cardCondition: card =>
+                    card.location === 'play area'
+                    && card.getType() === 'character'
+                    && card.kneeled
+                    && card.owner !== this.controller
```
Comment (ystros) [inline server/game/cards/19-JS/LeftHandLew.js:13]: You should add the missing `context` parameter to this card condition and use `context.player` instead.

File: server/game/cards/19-JS/LeftHandLew.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+            },
+            handler: context => {
+                context.target.controller.standCard(context.target);
+                this.game.addMessage('{0} kneels {1} to stand {2}', this.controller, this, context.target);
```
Comment (ystros) [inline server/game/cards/19-JS/LeftHandLew.js:17]: `context.player`.

File: server/game/cards/19-JS/MilkSnakes.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        });
+        this.action({
+            title: 'Move 1 gold to character',
+            condition: () => this.hasToken(Tokens.gold),
+            target: {
+                type: 'select',
+                cardCondition: card =>
+                    card.location === 'play area' &&
+                    card.getType() === 'character' &&
+                    card.controller === this.controller &&
```
Comment (ystros) [inline server/game/cards/19-JS/MilkSnakes.js:18]: `context.player`.

File: server/game/cards/19-JS/MilkSnakes.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+                    card.hasTrait('Clansman')
+            },
+            handler: context => {
+                this.game.transferGold({ from: this, to: context.target, amount: 1 });
+                this.game.addMessage('{0} moves 1 gold from {1} to {2}', this.controller, this, context.target);
```
Comment (ystros) [inline server/game/cards/19-JS/MilkSnakes.js:23]: `context.player`.

File: server/game/cards/19-JS/SerBorosBlount.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+const GameActions = require('../../GameActions');
+
+class SerBorosBlount extends DrawCard {
+    setupCardAbilities() {
+        this.reaction({
+            when: {
+                onCardEntersPlay: event => event.card === this
+            },
+            handler: context => {
+                this.game.promptForDeckSearch(this.controller, {
```
Comment (ystros) [inline server/game/cards/19-JS/SerBorosBlount.js:11]: `context.player`.

File: server/game/cards/19-JS/SerBorosBlount.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                    numCards: 10,
+                    activePromptTitle: 'Select a card',
+                    cardCondition: card => card.getType() === 'character' && (card.hasTrait('Kingsguard') || (card.hasTrait('Knight') && !card.isFaction('lannister'))) && this.controller.canPutIntoPlay(card),
```
Comment (ystros) [inline server/game/cards/19-JS/SerBorosBlount.js:14]: `context.player`.

File: server/game/cards/19-JS/SupportFromTheIronBank.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+                this.untilEndOfPhase(ability => ({
+                    effect: ability.effects.reduceCost({
+                        playingTypes: ['marshal', 'play'],
+                        amount: 1,
+                        match: card => card.isFaction('baratheon') && card.getType() !== 'character'
+                    })
+                }));
+
+                this.game.addMessage('{0} uses {1} to reduce the cost of each {2} non-character card they marshal or play this phase by 1',
+                    this.controller, this, 'baratheon');
```
Comment (ystros) [inline server/game/cards/19-JS/SupportFromTheIronBank.js:16]: `context.player`.

File: server/game/cards/19-JS/TheWinterMaid.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+            when: {
+                onPlotsRevealed: () => true
+            },
+            max: ability.limit.perPhase(1),
+            handler: context => {
+                let buttons = context.event.plots.map(plot => {
+                    return { method: 'plotSelected', card: plot, mapCard: true };
+                });
+
+                this.game.promptWithMenu(this.controller, this, {
```
Comment (ystros) [inline server/game/cards/19-JS/TheWinterMaid.js:15]: `context.player`.

File: server/game/cards/19-JS/Wraith.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+        });
+
+        this.action({
+            title: 'Kneel location',
+            cost: [
+                ability.costs.kneelSelf(),
+                ability.costs.sacrificeSelf()
+            ],
+            targets: {
+                location: {
```
Comment (ystros) [inline server/game/cards/19-JS/Wraith.js:17]: This nesting (`targets.location`) is unnecessary when only selecting one set of targets. It's fine, but just `target` would be simpler.

**Review by ystros — CHANGES_REQUESTED** [review]

File: server/game/player.js (hunk 1 above)
Comment (EvadLhorg) [reply server/game/player.js:164]: Yes, having additional used plots (without a trait) added by Ricasso (for example) and having an additional CITY used plot added by Gulltown should be two additional used plots in total.

File: server/game/player.js (hunk 2 above)
Comment (EvadLhorg) [reply server/game/player.js:240]: Might be we don´t need that event at all, i just copied it from the already present modifyUsedPlots method above without looking for references to the event. I will double check if it is needed to update the game state for cards referencing the used plot counter and post my findings

File: server/game/cards/19-JS/Brimstone.js (hunk 4 above)
Comment (EvadLhorg) [reply server/game/cards/19-JS/Brimstone.js:15]: good catch, will fix this

File: server/game/cards/19-JS/Brimstone.js (hunk 5 above)
Comment (EvadLhorg) [reply server/game/cards/19-JS/Brimstone.js:18]: will fix this

File: server/game/cards/19-JS/Wraith.js (hunk 18 above)
Comment (EvadLhorg) [reply server/game/cards/19-JS/Wraith.js:17]: good catch, fixed

File: server/game/player.js (hunk 2 above)
Comment (EvadLhorg) [reply server/game/player.js]: Tested the interaction between Host of the Boneway, Ricasso and Gulltown without this line of code (and the other one in the method above) and it didn´t change the behaviour. The Host´s STR was increased as expected.
=> i removed the two lines

**Comment by EvadLhorg** [comment]
Whoops, didn´t notice that you already updated the json-data reference. But i don´t think my latest commit messed anything up.
I changed the status of this PR from "draft" to "ready for review"

**Review by ystros — APPROVED** [review]

**Pull request merged**